### PR TITLE
Upgrade third_party/protobuf to 3.6.1

### PR DESCRIPTION
**Pull request by jtattermusch**
To make sure that updating to new protobuf version won't break us.

Things missing:
- updating the dependencies in languages that use protobuf packages (C#, python, ....)

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by jtattermusch** [comment]
FTR here are the changes we've done the last time we updated protobuf submodule: https://github.com/grpc/grpc/pull/15114

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
mac failures are caused by https://github.com/grpc/grpc/issues/15600, we need to re-run after fixing.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                                            cpu_time    real_time
-------------------------------------------------------------------  ----------  -----------
BM_PumpStreamServerToClient<InProcess>/32768                         +4%         +4%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/262144/262144  +5%         +5%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/262144    +8%         +8%
BM_Zalloc/512                                                        +4%         +4%
```

**Comment by acozzette** [comment]
It looks like "Artifact Build Linux" failed because it attempts to build with GCC 4.7.2, which does not have full support for C++11. Does gRPC still need to support that version?

**Comment by acozzette** [comment]
In some Python tests I also see this kind of error:
`TypeError: __new__() got an unexpected keyword argument 'serialized_options'`
I believe this indicates that the test must be mixing new generated code with an older runtime, which is not supported. In this [commit](https://github.com/google/protobuf/commit/0400cca3236de1ca303af38bf81eab332d042b7c) we added that serialized_options argument in `python/google/protobuf/descriptor.py`.

**Comment by jtattermusch** [comment]
@acozzette   ad the protoc artifact builds:
we are using and old linux distro to build protoc artifacts to make sure the resulting binaries are compatible with a large range of linux distros. We can change the way the protoc artifacts are built, but it's not completely trivial and requires some extra testing.
https://source.cloud.google.com/results/invocations/2d1acfd4-db7d-4ada-a30f-97398174461f/targets/github%2Fgrpc/tests

About mixing older runtime with new generated code: I think you are correct, but we can't really test the new generated code with new runtime in python without having access to the new python protobuf packages - so in python's case we'll probably need to take a leap of faith, as we don't have the release candidate packages right now.

There's also a nanopb failure:
```
++ basename src/proto/grpc/lb/v1/load_balancer.proto
+ protoc --plugin=protoc-gen-nanopb=/var/local/git/grpc/third_party/nanopb/generator/protoc-gen-nanopb '--nanopb_out=-T -L#include\ \"third_party/nanopb/pb.h\":/tmp/tmp.f0nV74llsl' load_balancer.proto
         ********************************************************************
         *** Failed to import the protocol definitions for generator.     ***
         *** You have to run 'make' in the nanopb/generator/proto folder. ***
         ********************************************************************
Traceback (most recent call last):
  File "/var/local/git/grpc/third_party/nanopb/generator/nanopb_generator.py", line 33, in <module>
    import proto.nanopb_pb2 as nanopb_pb2
  File "/var/local/git/grpc/third_party/nanopb/generator/proto/nanopb_pb2.py", line 26, in <module>
    dependencies=[google_dot_protobuf_dot_descriptor__pb2.DESCRIPTOR,])
TypeError: __new__() got an unexpected keyword argument 'serialized_options'
--nanopb_out: protoc-gen-nanopb: Plugin failed with status code 1.
```
https://source.cloud.google.com/results/invocations/a347bc9b-633d-46dc-8dd0-5af841752eee/targets/github%2Fgrpc%2Fsanity_linux_dbg_native/tests


Besides this, I see a distribtest failure, but that one can be relatively easily fixed by upgrading the cmake version in our dockerimage
https://source.cloud.google.com/results/invocations/2345720c-f8f8-4711-ae12-a1520766891b/log

**Comment by jtattermusch** [comment]
@acozzette   if we come up with a good explanation for the nanopb problem mentioned above, I think I should be relatively safe to tag the 3.6.0 release for protobuf and release the protobuf packages for all languages.  I can't guarantee everything will go smooth, but we should be able to bump our third_party/protobuf submodule to 3.6.0 without too many problems.

**Comment by jtattermusch** [comment]
The failed windows distribtest raises some concern if protobuf builds well with on Windows with VS2015  https://source.cloud.google.com/results/invocations/501a62be-2043-4aa3-81bb-ec499e43b763/targets/github%2Fgrpc/tests

**Comment by acozzette** [comment]
@jtattermusch Thanks for all the info! That nanopb problem looks like an instance of the same problem with the other Python tests (mixing new generated code with old runtime), so that may be OK. I'll look at the Windows error in just a minute.

**Comment by acozzette** [comment]
@jtattermusch I haven't been able to figure out what caused the VS2015 build failure. It's strange because it shows a bunch of errors inside system headers like <cmath>. That makes me think something must be wrong with the C++11 setup. Maybe the build is somehow mixing C++98 and C++11 code in an incompatible way?

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                      cpu_time    real_time
---------------------------------------------  ----------  -----------
BM_PumpStreamClientToServer<InProcess>/262144  +4%         +4%
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
This is currently being blocked by https://github.com/google/protobuf/issues/4874.

**Comment by jtattermusch** [comment]
we should align the changes to protoc artifact build Dockerfile with https://github.com/google/protobuf/pull/4877.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
The Windows build problem was resolved by https://github.com/google/protobuf/pull/4878.

**Comment by dgquintas** [comment]
Broken sanity: https://source.cloud.google.com/results/invocations/5646354a-5ea1-45f4-840a-636768778b02/targets/github%2Fgrpc%2Fsanity_linux_dbg_native/tests

**Comment by bobbytables** [comment]
Is there anything I can do to help move this along?

**Comment by Brahmasmi** [comment]
Protobuf 3.6.1 seems to have been released - https://github.com/google/protobuf/releases.

Thanks.

**Comment by jtattermusch** [comment]
Upgraded this PR to protobuf 3.6.1 (but python and C# nuget packages for protobuf 3.6.1 haven't been released yet, so they stay at 3.6.0 for now - let's see if that breaks any tests)

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: CMakeLists.txt [hunk 1]
```diff
@@ -100,6 +100,8 @@ if (MSVC)
   add_definitions(-D_WIN32_WINNT=0x600 -D_SCL_SECURE_NO_WARNINGS -D_CRT_SECURE_NO_WARNINGS -D_WINSOCK_DEPRECATED_NO_WARNINGS)
   # needed to compile protobuf
   add_definitions(/wd4065 /wd4506)
+  # TODO(jtattermusch): revisit warnings disabled when upgrading to protobuf3.6.0
```
Comment (nicolasnoble) [inline CMakeLists.txt]: When upgrading to 3.6? Isn't it what this PR is about?

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: CMakeLists.txt (hunk 1 above)
Comment (jtattermusch) [reply CMakeLists.txt]: Rephrased.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/16270

**Comment by jtattermusch** [comment]
@muxi  is there any ObjC changes needed for this pull request?
I see the iOS Binary Size Diff failing:
https://source.cloud.google.com/results/invocations/c126880a-abb9-4af7-83f5-5bab883f290c/targets/grpc%2Fcore%2Fpull_request%2Fmacos%2Fgrpc_ios_binary_size/log

```
Installing !ProtoCompiler (3.5.0)
[!] /bin/bash -c
set -e
if [ ! -f bin/protoc ]; then
  cd ../..
  # This will build protoc from the Protobuf submodule of gRPC, and put it in
  # ../../bins/opt/protobuf.
  #
  # TODO(jcanizales): Make won't build protoc from sources if one's locally installed, which
  # _we do not want_. Find a way for this to always build from source.
  make grpc_objective_c_plugin
  cd -
else
  mv bin/protoc .
  mv include/google .
fi
make: *** No rule to make target `third_party/protobuf/src/google/protobuf/inlined_string_field.h', needed by `/Volumes/BuildData/tmpfs/src/github/grpc/objs/opt/src/compiler/cpp_generator.o'.  Stop.
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
   412,978   Protobuf (<)        442,962

 1,950,829      Total (<)      1,981,730

***************FRAMEWORKS****************
  New size                      Old size
 1,859,645   Protobuf (<)      2,004,725

10,668,839      Total (<)     10,814,016


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/16270
https://github.com/grpc/grpc/issues/16280

**Pull request merged**